tests/lib/state_test.py

```diff
@@ -138,3 +140,122 @@ def _extra_state(self):
+        Route.NEXT_HOP_INTERFACE: IPV4_ROUTE1_IFACE_NAME,
+        Route.TABLE_ID: 254
+    }
+
+    IPV4_ROUTE2_IFACE_NAME = 'eth2'
+
+    IPV4_ROUTE2 = {
+        Route.DESTINATION: '203.0.113.0/24',
+        Route.METRIC: 102,
+        Route.NEXT_HOP_ADDRESS: '192.0.2.2',
+        Route.NEXT_HOP_INTERFACE: IPV4_ROUTE2_IFACE_NAME,
+        Route.TABLE_ID: 254
+    }
+
+    def test_merge_route_add(self):
+        """
+        Test adding routes with old routes preserved.
+        """
+        current_state = state.State(self._base_state_dict)
+        desired_state_dict = copy.deepcopy(self._base_state_dict)
+        new_route = copy.deepcopy(TestAssertRouteState.IPV4_ROUTE1)
+        new_route[Route.NEXT_HOP_ADDRESS] = '192.0.2.3'
+        desired_state_dict[Route.KEY][Route.CONFIG] = [new_route]
+        desired_state = state.State(desired_state_dict)
+        desired_state.merge_route_config(current_state)
+
+        assert new_route in desired_state.iface_routes['eth1']
+        iface_routes = desired_state.iface_routes
+        # Old routes should be no touched
+        assert (TestAssertRouteState.IPV4_ROUTE1 in
+                iface_routes[TestAssertRouteState.IPV4_ROUTE1_IFACE_NAME])
+        # Interface without changes should not be included
+        assert (TestAssertRouteState.IPV4_ROUTE2 not in
+                iface_routes[TestAssertRouteState.IPV4_ROUTE2_IFACE_NAME])
+
+    def test_merge_route_add_without_iface(self):
+        """
+        Test desired_state only contains routes without any interface defined.
+        """
+        current_state = state.State(self._base_state_dict)
+        desired_state_dict = copy.deepcopy(self._base_state_dict)
+        new_route = copy.deepcopy(TestAssertRouteState.IPV4_ROUTE1)
+        new_route[Route.NEXT_HOP_ADDRESS] = '192.0.2.3'
+        desired_state_dict[Route.KEY][Route.CONFIG] = [new_route]
+        desired_state_dict[Interface.KEY] = {}
+        desired_state = state.State(desired_state_dict)
+        changed_iface_names = desired_state.merge_route_config(current_state)
+        desired_state.include_changed_interfaces(current_state,
+                                                 changed_iface_names)
+
+        assert (new_route[Route.NEXT_HOP_INTERFACE] in
+                desired_state.interfaces)
+
+    def test_verify_routes_add_route(self):
+        """
+        Test route verification with the ordering of routes
+        """
+        current_state = state.State(self._base_state_dict)
+        desired_state_dict = copy.deepcopy(self._base_state_dict)
+        tmp_route = desired_state_dict[Route.KEY][Route.CONFIG][0]
+        desired_state_dict[Route.KEY][Route.CONFIG][0] = \
+            desired_state_dict[Route.KEY][Route.CONFIG][1]
+        desired_state_dict[Route.KEY][Route.CONFIG][1] = tmp_route
+        desired_state = state.State(desired_state_dict)
+        desired_state.verify_routes(current_state)
+
+    def test_add_route_without_metric(self):
```

Comment:
I do not understand what happens here.
To create the new state, a route is copied, its metric field is removed, interfaces subtree removed (why?), then after the merge, a default metric is added to the new route (in practice overwriting what was there before, embedded in the new state object). Finally the new route is checked to exist in the destination.

You better check what exactly exists in the new state after the merge, perhaps something got overwritten/removed.
There is also a matter of checking against the general state (s.state) and not only the indexed output (relevant to all tests).